Zulip as OAuth provider

edith007 opened the pull request:
This PR is the continuation of the work already done by @showell and @aero31aero in the following PR:
https://github.com/zulip/zulip/pull/16529

The aim is to enable Zulip to work as an OAuth provider. This is the issue: https://github.com/zulip/zulip/issues/17042

- [x] Refine read/write scopes
- [x] complete test coverage
- [x] add functionality to display errors made during form filling on registration page
- [x] Lock the views and only allow admin to create applications
- [ ] Swap the models to give the migrations a meaningful name. (ERROR with migration name -- 0002_auto_20190406_1805)  
- [ ] Move the views to the `settings page` of the Zulip application.
- [ ]  Investigate more and Add more in Zulip OAuth scopes (Currently - Read, Write)
- [ ] Build a configuration UI (like the Bots UI).


https://user-images.githubusercontent.com/53316982/113919646-762d4680-9801-11eb-8233-84db7e2c6df3.mp4


https://user-images.githubusercontent.com/53316982/113920927-0d46ce00-9803-11eb-98cf-026fca9fb249.mp4

showell committed:
Feature: Make Zulip an OAuth provider.

Enable Zulip to serve as an OAuth provider by using the
django-oauth-toolkit. We have locked down the views and we allow only
the admin to create applications. The toolkit is installed
automatically.
Add a test to verify the complete flow for the oauth authentication.

aero31aero commented on static/styles/portico/portico_signin.css:
```diff
@@ -927,6 +927,20 @@ button#register_auth_button_gitlab {
     font-size: 22px;
 }
 
+#clint_type_section {
```
typo; should be `client_type_section`.

aero31aero commented on static/styles/portico/portico_signin.css:
```diff
@@ -927,6 +927,20 @@ button#register_auth_button_gitlab {
+    font-size: 22px;
+}
+
+#authorization_grant_type_section {
```
Merge these two selectors:

```css
#id-1, #id-2 {
    /* rules */
}
```

Better yet, give both of them a class like `oauth_section`.

aero31aero commented on static/styles/portico/portico_signin.css:
```diff
@@ -936,6 +950,24 @@ button#register_auth_button_gitlab {
+    }
+}
+
+#authorization_grant_type_select {
```
Same logic: give them both a class `oauth_select`.

aero31aero commented on templates/oauth2_provider/application_confirm_delete.html:
```diff
@@ -0,0 +1,23 @@
+{% extends "oauth2_provider/base.html" %}
+
+{% block portico_content %}
+<div class="register-account flex full-page">
```
Forgot to remove the `register-account` class?

aero31aero commented on templates/oauth2_provider/application_confirm_delete.html:
```diff
@@ -0,0 +1,23 @@
+            <h3 class="block-center-heading">{{_("Are you sure to delete the application")}} {{ application.name }}?</h3>
+
+            <div class="block-center">
+                <a class="register-button" href="{{ url('oauth2_provider:list') }}">{{ _("Cancel") }}</a>
```
Similarly, this and the next line mention `register-button`

aero31aero reviewed (commented):
I took a partial look at the second commit and it looks like you used register page as the base here:

```
$ git show HEAD | grep register
@@ -927,6 +927,20 @@ button#register_auth_button_gitlab {
@@ -936,6 +950,24 @@ button#register_auth_button_gitlab {
+<div class="register-account flex full-page">
+                <a class="register-button" href="{{ url('oauth2_provider:list') }}">{{ _("Cancel") }}</a>
+                <button type="submit" class="register-button" value="Delete">{{ _("Delete") }}</button>
+<div class="register-account flex full-page">
+            <a class="register-button" href="{{ url('oauth2_provider:list') }}">
+            <a class="register-button" href="{{ url('oauth2_provider:update', args=[application.id]) }}">
+            <a class="register-button" href="{{ url('oauth2_provider:delete', args=[application.id]) }}">
+<div class="register-account flex full-page">
+            <div class="register-button-box">
+                <button class="register-button" type="submit">{{ _('Register') }}</button>
+<div class="register-account flex full-page">
+            <a href="{{ 'register' }}">
+                <div class="register-button-box">
+                    <button class="register-button" type="submit">{{ _('Create New Applcation') }}</button>
+{% block app_form_action_url %}{{ url('oauth2_provider:register') }}{% endblock %}
+<div class="register-account flex full-page">

```

I'd suggest running the above command and some manual inspection to make sure traces of this copying are removed.

Overall this is looking neat!

edith007 committed:
OAuth: Add CSS to make oauth pages consistent with Zulip.

Currently we show the default django-oauth-toolkit pages on our OAuth
end-points. We want to customize them as per Zulip's design and this
commit adds CSS which can be used for that.

edith007 committed:
OAuth: Add template/oauth2_provider/base.html.

Add base.html which extends portico_signup.html to show Zulip's header
and footer across OAuth pages. All the other OAuth pages will extend
this.

edith007 committed:
OAuth: Add outh2_provider/authorize.html.

This page is rendered when the application consuming Zulip's OAuth
service wants to authorize its user.

edith007 committed:
OAuth: Add oauth2_provider/application_list.html.

This page contains all the OAuth applications created by a user.

edith007 committed:
OAuth: Add oauth2_provider/application_registration_form.html.

This page uses registration_form.html and is rendered when a user wants
to register an OAuth application.

edith007 committed:
OAuth: Add oauth2_provider/application_form.html.

This page renders the application form used when a user either registers
a new application or edits an existing one.

edith007 committed:
OAuth: Add oauth2_provider/application_detail.html.

This page renders the detail of a particular OAuth application created
by a user. This page also provides the option to edit or delete the
application.

edith007 committed:
OAuth: Add oauth2_provider/application_confirm_delete.html.

This page is rendered when a user wants to delete an OAuth application.
This page simply asks the user to confirm the deletion.

edith007 committed:
OAuth: Add oauth2_provider/authorized-tokens.html.

This page gets a list of authorized tokens.

edith007 committed:
OAuth: Add oauth2_provider/authorized-token-delete.html.

This page is rendered when a token has to be deleted.

edith007 committed:
decorators: Remove stale exception handler in API authentication.

This commit removes the stale exception handler. Earlier we only had to
look for HTTP_AUTHORIZATION header in the request, so it made sense to
have a try block and catch for the KeyError exception. Now, with the
addition of OAuth we also check for HTTP_BEARER header in the request,
so we check for the headers using if-elif-else ladder. So, the control
will never go the exception handler for KeyError. Hence, this commit
removes that stale code.

edith007 committed:
refactor: Break the OAuth flow in functions.

This commit breaks the steps in the OAuth flow and places them in their
respective functions. This enables us to write the tests easily and not
to repeat the same code in every test.

The flow for OAuth is as follows:

- Create an OAuth application, which is done in setUp function

- Then we start the OAuth code flow by visiting the url /o/authorize/,
which is done in start_oauth_code_flow function

- Then we are asked to authorize the consumer to access the resources,
this is taken care of in the authorize_the_web_app function. At this
point, the OAuth application also gives a code to the consumer. This
function also takes care of that.

- Then the consumer gets a refresh token which it can use to access the
resources, this is done in the function called get_access_token. This
function returns the access_token.

Finally we use this access_token as the HTTP_BEARER header in our
requests to the zulip server.

edith007 committed:
OAuthTest: Test for invalid access token.

This commit adds a test to check if the authentications fails when an
incorrect access token is passed as the HTTP_BEARER header.

edith007 committed:
OAuthTest: Test for not supporting incoming webhook bots.

We do not support OAuth for webhooks. So, if a webhook makes a request,
we raise a JsonableError. This test checks for that.

edith007 committed:
OAuth: Add API endpoint for OAuth Creation.

This endpoint will allow client to create a OAuth application containing
the user_id, application_name, client_id, client_secret, and
redirect_uri for their OAuth Application.

edith007 committed:
oauth_application:Add UI to list in oauth settings page.
